### PR TITLE
Remove capnproto

### PR DESCRIPTION
**Pull request by dkeeney**
This is still not expected to compile.  There are still some errors in the Swig generated code I am trying to resolve.

### CODE REVIEW

**Comment by dkeeney** [comment]
I still have one problem:  Swig:  "Error: Nothing known about namespace 'nupic::algorithms::connections'
while compiling algorithms.i.  It does have %include <nupic::algorithms::Connections.hpp> which defines the namespace 'nupic::algorithms::connections' but Swig is still very confused about something.

I don't really want to spend very much more time debugging code (the SWIG bindings) which we will be deleting entirely.  Can someone take a quick look and see if they can find the problem. If not, shall I just go on to the next PR and delete SWIG?  This does compile on Windows without SWIG.

File: src/nupic/algorithms/Cells4.cpp [hunk 1]
```diff
@@ -2183,16 +1954,14 @@ void Cells4::save(std::ostream &outStream) const {
...
-  outStream.precision(std::numeric_limits<double>::digits10 + 1);
-  save(outStream);
-
-  // Explicitly close the stream so that we may get an exception on error
-  outStream.close();
+//do later  std::filesystem::path p(filePath.c_str());
+//do later  std::filesystem::create_directories(p.parent_path());
+  std::ofstream out(filePath.c_str(),std::ios_base::out | std::ios_base::binary);
+  out.exceptions(std::ofstream::failbit | std::ofstream::badbit);
+  out.precision(std::numeric_limits<double>::digits10 + 1);
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/algorithms/Cells4.cpp]: I have no idea what the +1 is for.   All of the examples showed using it
f << std::setprecision(std::numeric_limits<long double>::digits10 + 1);
The docs don't say what it is for.

File: src/nupic/algorithms/Cells4.hpp [hunk 2]
```diff
@@ -637,510 +645,513 @@ class Cells4 : public Serializable<Cells4Proto> {
...
+
+        /**
+         * TODO: compute, rather than writing to a buffer.
+         * TODO: move persistence to binary, faster and easier to compute expecte size.
+         */
+        UInt persistentSize() const
+        {
+          // TODO: this won't scale!
+          std::stringstream tmp;
+          this->save(tmp);
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/algorithms/Cells4.hpp]: Don't think so.  Just 64 bit address space.  It would swap onto disk if it got too big but it would still work.

File: src/nupic/types/Serializable.hpp [hunk 3]
```diff
@@ -22,48 +22,72 @@
 
 /** @file
  * Definitions for the base Serializable class in C++
+ *
+ * CapnProto serialization has been removed and replaced with binary streams.
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: Good idea

**Comment by dkeeney** [comment]
I have this branch building again.  The problem was that I was too aggressive when ripping things out of algorithm.i and got a few more lines that I intended.  That was a tough one to debug.
I will upload it as soon as I address the rest of breznak's comments on the last one.

**Comment by dkeeney** [comment]
Another design question.  We have
- serialization to binary   (the save() and load())
- serialization to text
- toString()
Which one should the << and >> overloading apply to.  I am thinking it should be serialization to text. Does anyone care?

File: src/nupic/types/Serializable.hpp [hunk 4]
```diff
@@ -22,48 +22,72 @@
...
 
 #include <iostream>
+#include <fstream>
+#include <nupic/os/Directory.hpp>
+#include <nupic/os/Path.hpp>
 
-#include <capnp/message.h>
-#include <capnp/serialize.h>
-#include <kj/std/iostream.h>
+#define SERIALIZABLE_VERSION 1
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: There is a problem with doing that...some of the things being serialized depend on the version of the object not the version of the overall Serialization protocol.

File: src/nupic/types/Serializable.hpp [hunk 5]
```diff
@@ -22,48 +22,72 @@
...
+  virtual void load(std::istream &stream) = 0;
 
   virtual ~Serializable() {}
 };
 
 } // end namespace nupic
-#endif // NTA_serializable_HPP
+
+/*****
+using namespace nupic::serializable;
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: That is what I was thinking.  However it means that anyone trying to use this must also include yet another namespace.  The objective of a namespace is to avoid name collisions.  Since the Serializable class can appear anywhere within the nupic namespace and save() and load() should not mean anything else, it seems reasonable to include it in the nupic namespace.  So I am leaving out the serializable namespace.

File: src/test/unit/algorithms/Cells4Test.cpp [hunk 6]
```diff
@@ -179,94 +175,7 @@ TEST(Cells4Test, pickleSerialization) {
   }
 }
 
-TEST(Cells4Test, capnpSerialization) {
-  Cells4 cells(10, 2, 1, 1, 1, 1, 0.5, 0.8, 1, 0.1, 0.1, 0, false, -1, true,
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/test/unit/algorithms/Cells4Test.cpp]: Note the test just before this one is pickleSerialization. This is the same test but using binary serialization.  I renamed it just Serialization because it is pickling only in the Python.

File: src/nupic/types/Serializable.hpp [hunk 7]
```diff
@@ -22,48 +22,72 @@
...
+    std::ifstream in(filePath.c_str(), std::ios_base::in | std::ios_base::binary);
+    in.exceptions(std::ifstream::failbit | std::ifstream::badbit);
+    load(in);
+    in.close();
   }
 
-  virtual void write(typename ProtoT::Builder &proto) const = 0;
-  virtual void read(typename ProtoT::Reader &proto) = 0;
+  // These must be implemented by the subclass.
+  virtual void save(std::ostream &stream) const = 0;
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: Fixed.  Good catch

**Comment by breznak** [comment]
Dave, great you've managed to fix the stuff in Algorithms,i ! 
I'm at a conference this week, so apologies for slower replies.. 

> Which one should the << and >> overloading apply to. I am thinking it should be serialization to text. Does anyone care? 

by "serialization to text" you mean a nicer print()? ie python's `__str__` meaning? Otherwise, if there were 2 full-blown serializations, I'd go with just one (textual)

File: src/nupic/algorithms/Svm.hpp [hunk 8]
```diff
@@ -325,7 +326,8 @@ struct decision_function {
  * n_sv = n_sv[n_class], number of SVs for each class
  * probA, probB = [n_class*(n_class-1)/2]
  */
-class svm_model {
+class svm_model  : public Serializable
```
Comment (breznak) [inline src/nupic/algorithms/Svm.hpp]: btw, we should investigate if we can get rid off this file

File: src/nupic/engine/Link.cpp [hunk 9]
```diff
@@ -476,7 +474,7 @@ void Link::serialize(std::ostream &f) {
     Array a = dest_->getData().subset(destOffset_, srcCount);
     f << a; // our part of the current Dest Input buffer.
 
-    std::deque<Array>::iterator itr;
+    boost::circular_buffer<Array>::iterator itr;
```
Comment (breznak) [inline src/nupic/engine/Link.cpp]: why this change from std:: ?

File: src/nupic/math/Array2D.hpp [hunk 10]
```diff
@@ -36,6 +36,8 @@
 // try to separate algorithms and write them for all types of matrices??
 // optimize kernels depending on cache sizes?
 
+// Note: cannot be made a nupic::Serializable class because it is not in nupic namespace.
```
Comment (breznak) [inline src/nupic/math/Array2D.hpp]: can you add it to ns nupic:: ?

File: src/nupic/math/ArrayAlgo.hpp [hunk 11]
```diff
@@ -24,4952 +24,6021 @@
  * Algorithms on arrays, dense or sparse. Contains speed-optimized asm
  * code that uses SIMD instructions provided by SSE in x86, only for darwin86
  * so far.
+ *
+ * Notes:
+ *	bind2nd() is depriciated in C++11 and removed in C++17 so
+ *            bind2nd is replaced with C++11 bind(). At some point we may
+ *            want to consider using C++14 generic lambda which they say is faster.
+ *  Compiles under Windows C++17 and Linux C++11.  Asm no used in Windows.
```
Comment (breznak) [inline src/nupic/math/ArrayAlgo.hpp]: the comment and change to bind() is :+1: 
but this file is probably a merge conflict! Too many changes here, 
please see the original 
https://github.com/htm-community/nupic.cpp/blob/master/src/nupic/math/ArrayAlgo.hpp 
And apply your change to that. 
I think you've re-added the ASM stuff here (which might be the reason why CI builds are failing) 
((also, this removed ASM is another thing to add to the "Changes", although it's not API(?) )

File: src/nupic/math/StlIo.hpp [hunk 12]
```diff
@@ -551,37 +551,48 @@ inline std::ostream &operator<<(std::ostream &out_stream,
 
 //--------------------------------------------------------------------------------
 // std::map
+// Warning: This will not handle elements containing whitespace.
```
Comment (breznak) [inline src/nupic/math/StlIo.hpp]: can we NTA_ASSERT on this?

File: src/nupic/ntypes/ArrayBase.cpp [hunk 13]
```diff
@@ -391,138 +392,7 @@ void ArrayBase::binaryLoad(std::istream &inStream) {
...
-  ////////////////////////////////////////////////////////////////////////////////
 
-  void ArrayBase::serialize(YAML::Emitter& out) const
-  {
-    out << YAML::BeginMap;
-    out << YAML::Key << "type" << YAML::Value  << BasicType::getName(type_);
-    out << YAML::Key << "count" << YAML::Value << count_;
-    out << YAML::Key << "buffer" << YAML::Value << YAML::BeginSeq;
-    switch (type_) {
-    case NTA_BasicType_Byte:
```
Comment (breznak) [inline src/nupic/ntypes/ArrayBase.cpp]: nice cleanup!

File: API_CHANGELOG.md [hunk 14]
```diff
@@ -0,0 +1,7 @@
+Breaking changes to the nupic API
+
+* CapnProto serialization is replaced with binary streams in PR #62.  
+Calls to read() and write() are no longer available. Use save() and load(). Network(path) is no 
+longer used to deserialize a path. Use Network net; net.load(stream); to deserialize.  
+Helpers SaveToFile(path) and LoadFromFile(path) are used to stream to and from a file using save()
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply API_CHANGELOG.md]: ok

File: src/nupic/algorithms/Cells4.cpp [hunk 15]
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
   //}
 }
 
+struct self_t {
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/algorithms/Cells4.cpp]: Oh, I forgot about compiler generated padding.  Good catch.
I did zero fill the structure before filling it so any padding will be nulls.

File: src/nupic/algorithms/Cells4.cpp [hunk 16]
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
...
+  Real avgLearnedSeqLength;
+  UInt maxAge;
+  UInt verbosity;
+  Int maxSegmentsPerCell;
+  Int maxSynapsesPerSegment;
+  Real avgInputDensity;
+  UInt pamCounter;
+};
+// NOTE:  ownsMemory cannot be saved in serialization.
+//        It indicates who currently allocated the buffers.
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/algorithms/Cells4.cpp]: Well...I need this for CSharp too because CSharp has a different heap (managed) than C++.  It cannot directly access C++'s heap space.  So to avoid a copy we need to do this.  I will try to make this a little cleaner however.

File: src/nupic/algorithms/Cells4.cpp [hunk 17]
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
...
-  outStream << _learnActiveStateT << " " << _learnActiveStateT1 << " "
-            << _learnPredictedStateT << " " << _learnPredictedStateT1
-            << std::endl;
-
-  // Additions in version 2.
-  outStream << _segmentUpdates.size() << " ";
+
+  // Capture the class variables and write them as a binary block.
+  struct self_t self;
+  memset(&self, 0, sizeof(self));
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/algorithms/Cells4.cpp]: Its my old C roots that are showing :)
However, initializing with {} does a field by field initialization.  That is not what I am looking for. I want to zero any compiler generated padding and the {} would not do it.  So I think it should remain memset(). I will comment that this structure must contain only POD (plain old data).

File: src/nupic/algorithms/Cells4.cpp [hunk 18]
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
...
-  load(outStream);
+  outStream << "out " << std::endl;
 }
 
 //------------------------------------------------------------------------------
 /**
  * Need to load and re-propagate activities so that we can really persist
  * at any point, load back and resume inference at exactly the same point.
+ *
+ * Note: Most parts are saved in binary format. The file
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/algorithms/Cells4.cpp]: It is not the size that is important it is the enormous time required to format/parse to and from text.  Even for a small implementation the time is significant.

File: src/nupic/engine/Link.cpp (hunk 9 above)
Comment (dkeeney) [reply src/nupic/engine/Link.cpp]: I started to make that a std::deque but discovered that it breaks other code that should be in a different PR.  It's part of a PR to remove dimensions which ends up removing LinkPolicy and removes a lot of complexity in the Link object.  So lets delay that for now.

File: src/nupic/math/Array2D.hpp (hunk 10 above)
Comment (dkeeney) [reply src/nupic/math/Array2D.hpp]: Well...perhaps but lets make that a separate PR.  It could get messy.

File: src/nupic/math/ArrayAlgo.hpp (hunk 11 above)
Comment (dkeeney) [reply src/nupic/math/ArrayAlgo.hpp]: This is a tough one.  I am trying to find something that will compile under C++11 and C++17.  I ended up getting really messed up such that nothing compiled.  So I went back to the original...perhaps I went back too far.  It now compiles.  I don't think the ASM stuff is all that bad. This was not the problem with CI.  It will not work on Windows but in those cases where it matters it does make a big difference in speed.   But I will work on this some more and reduce the conflicts.

File: src/nupic/math/StlIo.hpp (hunk 12 above)
Comment (dkeeney) [reply src/nupic/math/StlIo.hpp]: I am not sure this is actually used anywhere...just noticed this was the case while looking to add something like it for std::deque.   The way to allow whitespace would be to use readline and terminate with \n.  We can keep that in mind if we find we are using this function.

File: src/nupic/ntypes/ArrayBase.cpp (hunk 13 above)
Comment (dkeeney) [reply src/nupic/ntypes/ArrayBase.cpp]: :)

**Comment by dkeeney** [comment]
@breznak Thanks for taking time to review this for me.  It is a big help.

File: src/nupic/algorithms/Cells4.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (breznak) [reply src/nupic/algorithms/Cells4.cpp]: > I want to zero any compiler generated padding and the {} would not do it 

I'm interested why the "padding" bits would matter? As we are, so far, compiler&platform dependant.

File: src/nupic/algorithms/Cells4.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (breznak) [reply src/nupic/algorithms/Cells4.cpp]: ...understood, and pity

File: src/nupic/math/ArrayAlgo.hpp (hunk 11 above)
Comment (breznak) [reply src/nupic/math/ArrayAlgo.hpp]: > compile under C++11 and C++17 

for other PR, but.. you've convinced me to go c++17 way (but there were some problems with python you've later discovered?), so if there are problems between ++11/17, should we move to the middleman c++14 as a sub-goal? 

> where it matters it does make a big difference in speed 

Do you have a test for this? Because I was deciding just that, and in my evaluation the plain c++(with compiler optimizations) made no difference compared to the ASM - therefore I went on pruning all of the ASM. 

Anyway, my comment was that you got messed in rebasing from your other branch to the newer base, if there's not too much changes, I'd suggest go with the current file from this repo and rewrite your changes there manually.

Comment (dkeeney) [reply src/nupic/math/ArrayAlgo.hpp]: Ok, re-did this file.  Went back to nupic.cpp master, and re-made just my changes. .. not many.  Compiles on C++11 and C++17.

**Comment by dkeeney** [comment]
Missed updating all of the Py tests in bindings/py/ to remove capnproto.  This is going to take a while.

**Comment by breznak** [comment]
Thanks for your work, @dkeeney , this will be a really useful cleanup! 

I've started reading on upcoming parallelism in c++17, so I'm looking forward to that shift afterwards

**Comment by dkeeney** [comment]
Thanks for looking at this.
Yes, I am still trying to figure out what is generated and what is hand
coded in the nupic/bindings/py folder.  I would think that generated stuff
should be moved to the build folder so it does not get checked in....or at
least tell git not to include it.

I will probably not get much done until after the laber day weekend...big
family thing coming up.  But I will continue to work on this.

**Comment by breznak** [comment]
> Yes, I am still trying to figure out what is generated and what is hand coded in the nupic/bindings/py folder 

is your .gitignore in order? Mine does not check-in any of that stuff
```
...
# SWIG
bindings/py/src/nupic/bindings/*.cxx
bindings/py/src/nupic/bindings/*.py
bindings/py/src/nupic/bindings/*.so
bindings/py/src/nupic/*.so
bindings/py/src/nupic.bindings.egg-info
```
I think only the `bindings/*.i` files are relevant. 

Not sure about moving it to build/, anyway, let's leave that for separate PR.. 

Enjoy your time with family!

**Comment by breznak** [comment]
> There are lots of tests for capnproto style serialization but nothing for stream serialization so I will have to write some new tests. Have to think how this will tie into pickling of if it needs to.  

If it's pytests for serialization you speak of, maybe we could drop them? And keep mostly only c++ tests for serialization (+ extend them for some extra logic/testcases if found in PY?)

**Comment by breznak** [comment]
Btw, 

> The bindings/*.i is under nupic.cpp/src/nupic/bindings so that is a different bindings. 

this is where the manual code is, 

> These in nupic.cpp/bindings/py seem to be where the py tests reside. I had overlooked them entirely.

I think all of this is created by SWIG during build, thus can be ignored

File: src/nupic/bindings/algorithms.i [hunk 19]
```diff
@@ -1337,31 +1337,31 @@ void forceRetentionOfImageSensorLiteLibrary(void) {
...
+
+    // def convertedRead(self, proto):
+    //   """Initialize the Connections instance from the given ConnectionsProto
+    //   reader.
+    //
+    //  :param proto: ConnectionsProto message reader containing data from a
+    //                previously serialized Connections instance.
+
+    //  """
+    //  self._initFromCapnpPyBytes(proto.as_builder().to_bytes()) # copy * 2
```
Comment (breznak) [inline src/nupic/bindings/algorithms.i]: remove this commented code? If it were ever needed, we have git history

File: bindings/py/src/nupic/bindings/regions/PyRegion.py [hunk 20]
```diff
@@ -324,67 +326,7 @@ def deSerializeExtraData(self, filePath):
...
-    on subclass after converting proto to specific type using
-    :meth:`~nupic.bindings.regions.PyRegion.PyRegion.getSchema`.
-
-    :param proto: PyRegionProto capnproto object
-    """
-    regionImpl = proto.regionImpl.as_struct(self.getSchema())
-    self.writeToProto(regionImpl)
-
-
   @classmethod
```
Comment (breznak) [inline bindings/py/src/nupic/bindings/regions/PyRegion.py]: remove this line as well

File: bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py [hunk 21]
```diff
@@ -69,41 +58,8 @@ def randomSeed(self):
...
-  def writeToProto(self, proto):
-    """Write state to proto object.
-
-    The type of proto is determined by getSchema().
-    """
-    proto.dataWidth = self._dataWidth
-    self._rand.write(proto.random)
-
 
   @classmethod
```
Comment (breznak) [inline bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py]: remove this line

File: bindings/py/tests/nupic_random_test.py [hunk 22]
```diff
@@ -44,21 +36,17 @@
 class TestNupicRandom(unittest.TestCase):
 
 
-  @unittest.skipUnless(
-    capnp, "pycapnp is not installed, skipping serialization test.")
```
Comment (breznak) [inline bindings/py/tests/nupic_random_test.py]: keep the skipUnless logic? Could be nice to have a variable serialization (former capnp) and if set to None, the tests are skipped.

File: src/nupic/bindings/algorithms.i [hunk 23]
```diff
@@ -1337,31 +1337,31 @@ void forceRetentionOfImageSensorLiteLibrary(void) {
 
 
     @classmethod
```
Comment (breznak) [inline src/nupic/bindings/algorithms.i]: remove/comment out too

File: src/nupic/regions/PyRegion.cpp [hunk 24]
```diff
@@ -329,8 +329,17 @@ void PyRegion::serialize(BundleIO &bundle) {
 *  py::Ptr none1(node_.invoke("serializeExtraData", args1));
 ***********/
 	// pickle to a stream
-	//std::ostream &f = bundle.getOutputStream();
-	// @TODO
+	std::ostream &f = bundle.getOutputStream();
+	// cPickle.dump(node_, f, HIGHEST_PROTOCOL)
+	py::Module pickle("pickle")
```
Comment (breznak) [inline src/nupic/regions/PyRegion.cpp]: missing ';' ? 
https://travis-ci.org/htm-community/nupic.cpp/builds/422329818#L3876

File: src/nupic/bindings/algorithms.i (hunk 19 above)
Comment (dkeeney) [reply src/nupic/bindings/algorithms.i]: done

File: bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py (hunk 21 above)
Comment (dkeeney) [reply bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py]: No, executeMethod() that follows needs to be declared as a classmethod.  Otherwise I think it would be declared abstract.

File: src/nupic/regions/PyRegion.cpp (hunk 24 above)
Comment (dkeeney) [reply src/nupic/regions/PyRegion.cpp]: oops

File: bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py (hunk 21 above)
Comment (breznak) [reply bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py]: Do you reply to a good line/file? There's no executeMethod here. The decorator belonged to 'readFromProto', and now would fall down to 'initialize'

**Comment by dkeeney** [comment]
Hmmm, CI seems to be really broken now.  Don't know what I did that might have caused that.  My company is arriving in a few min so I will be out-of-pocket until Monday afternoon. I may be able to check my email but not much else.

**Comment by dkeeney** [comment]
Appveyor is still failing for some unknown reason.  Seems to be confused about the Serializable base class in the SWIG magic someplace.
The other two were successful but to do so I had to comment out the python tests.  The python tests do not seem to be able to load their modules.  The capnproto code had some magic that loaded the path. I cannot figure out how that should happen.  Anyway, all of this will be replaced with pybind11 so I don't think that is a real problem.

**Comment by dkeeney** [comment]
After looking at appveyor a little more, it says it has written out the maps and then it errors saying it cannot find the maps.  Don't know what to make of this other than that it is broken.
I have an idea of how I might get the python tests to work in the other two CI's.  I think I will give that at a try.

**Comment by dkeeney** [comment]
My idea did not work.  
There are actually two errors that stop the python tests from running. The first is some sort of undefined symbol:  _ZN5nupic6Region9intializeEv when loading nupic.bindings.engine_internal.   Probably something to do with nupic.Region.initialize() but there is nothing unusual about it that I can see.

The second error is an undefined name 'bindings' in class CState(bindings.math.Serializable):  in algorithms.py which is built by SWIG.  Serializable is indeed defined and it is in the math.so.  So I would expect SWIG to reference it as nupic.bindings.math.Serializable.   This is not just CState() but every class that subclasses Serializable.

So I am at a loss as to what to try next.

File: bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py (hunk 21 above)
Comment (dkeeney) [reply bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py]: Hmmm, I must have been looking at the wrong file.  Removed line 62 of serialization_test_py_region.py

**Comment by dkeeney** [comment]
I tried to update the src/CMakeLists.txt SWIG_ADD_MODULE to SWIG_ADD_LIBRARY.  That got rid of the warning but did not fix the problem.  The .expmap is still not generated under appveyor. The other two CI's have no problem.  Perhaps it is something about Windows wingw that is having a problem with it.

**Comment by breznak** [comment]
@dkeeney sorry, I just arrived home from a conference. I'm going to catch my sleep deficit and I'll have a look asap tmr! 

So we're considering breaking the Windows for now..

> The first is some sort of undefined symbol: _ZN5nupic6Region9intializeEv when loading nupic.bindings.engine_internal. Probably something to do with nupic.Region.initialize() but there is nothing unusual about it that I can see. 

There was a hack in capnp for adding a symbol, I remember seeing it being deleted. Maybe we'd have to hack it in.

**Comment by dkeeney** [comment]
Found a misspelled class method  Region::intialize() which was not implemented.  Appveyor seemed to have a problem resolving what Swig generated for that situation.

**Comment by dkeeney** [comment]
Well all three CI's now fail on the same Python test in the same way.  So Windows CI is working again.
I now need to figure out how to tell Swig how to generate code that will load the Serializable class.

**Comment by breznak** [comment]
>  So Windows CI is working again.
I now need to figure out how to tell Swig how to generate code that will load the Serializable class. 

This is good progress, so we have a single problem and even know the cause. Regarding the SWIG - I think we can find the approach from capnp that used Serializabe.

**Comment by dkeeney** [comment]
It mostly works :)   One of the Python tests crashed on appveyer. Will have to look at that in the morning.
I had a problem getting Swig to recognize the Serializable class.  I tried everything with no luck.  Ended up overriding the saveToFile() and loadFromFile() functions in the subclasses so it did not need the Serializable class exposed to Python.  

Also, had to comment out one of the Python test which performed serialization of the Python modules. That part not yet implemented. PyRegion.cpp line 293. That can be put off until the Python interface PR.

**Comment by dkeeney** [comment]
Well, I thought it might have been lack of permissions to write to a file (for serialization) but after finding a way to obtain a valid tmp directory from py.test it still fails.   :(  Not just a failed test but a hard crash of the test environment.  So I guess I need to install MinGW just to debug this problem.

**Comment by dkeeney** [comment]
Finally :)  
I think I have addressed all of the comments so far.
I am going to be out-of-packet for the next few days....one of the advantages of being retired.  I will check email but no more coding until next week.

For next project...perhaps something simple like adding Smart pointers to Region and Link objects.

**Comment by breznak** [comment]
Wow, this has been an amazing and a loong way, muchas gracias @dkeeney !! I'll review all of it again today and let's merge this beast :1st_place_medal: Enjoy your rest and bathe in sunshine of passing CI! :)

File: bindings/py/tests/network_test.py [hunk 25]
```diff
@@ -91,42 +83,43 @@ class NetworkTest(unittest.TestCase):
...
-      destNet = engine.Network.read(readerProto)
-
-      destRegion = destNet.getRegions().getByName(
-        SerializationTestPyRegion.__name__)
-
-      self.assertEqual(destRegion.getParameterUInt32("dataWidth"), 128)
-      self.assertEqual(destRegion.getParameterUInt32("randomSeed"), 99)
+#########################################
+#Skipping this test for now...
+#TODO: Need to implement pickle serialization of the Python code.
```
Comment (breznak) [inline bindings/py/tests/network_test.py]: this is still missing? Guess left over to pybind11 (?)

File: nupic.core.kdev4 [hunk 26]
```diff
@@ -0,0 +1,4 @@
+[Project]
+CreatedFrom=CMakeLists.txt
+Manager=KDevCMakeManager
```
Comment (breznak) [inline nupic.core.kdev4]: rm this file, please

File: src/nupic/algorithms/SpatialPooler.cpp [hunk 27]
```diff
@@ -32,14 +32,15 @@
 #include <nupic/algorithms/SpatialPooler.hpp>
 #include <nupic/math/Math.hpp>
 #include <nupic/math/Topology.hpp>
-#include <nupic/proto/SpatialPoolerProto.capnp.h>
+
+#define VERSION 2  // version for stream serialization
```
[the remark this answers is not in the document]
Comment (breznak) [reply src/nupic/algorithms/SpatialPooler.cpp]: ping

**Comment by breznak** [comment]
Merging capnp removal, replaced by native bitstream serialization! 
Huge work done by @dkeeney , thank you very much!! :clap: :clap:

File: .gitignore [hunk 28]
```diff
@@ -17,19 +18,29 @@ htmlcov/
 .vscode
 .ycm_extra_conf.py
 
+# KDevelop
+.kdev4
+nupic.core.kdev4
```
[the remark this answers is not in the document]
Comment (breznak) [reply .gitignore]: sorry, I just deleted that. Best solution would be to have the IDE to keep the files outside the git project's folder/scope, so individual settings don't have to be commited. Or just add it next time please.

File: nupic.core.kdev4 (hunk 26 above)
Comment (breznak) [reply nupic.core.kdev4]: probably you had the file first check to git, and later added to gitignore - it still stays in the repo. Just let the file removed here, and keep your local version - now it should not show in the public git, but you should keep your local config.
